Use package groups

philderbeast opened the pull request:
We have a lot of projects and they each have packages. Borrowing an idea from Updo, we could group packages and import these groups, fixes #9622. I've also:

- Removed the coverage and weeder projects.
- Except for the `meta` project, projects import package groups.
- The `release` project now builds with `--index-state="hackage.haskell.org HEAD` with `ghc-9.8.1` (#9600 does more in relation to this). A spin off from this work is the CI check of `cabal.project.release` in #9601.
- Projects that have additional constraints, all import the same constraints.
- Projects that set `ghc-options`, all import the same `ghc-options`.
- Projects that need `ghc-latest`, all import the same config for this.

**Template Β: This PR does not modify `cabal` behaviour (documentation, tests, refactoring, etc.)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

philderbeast commented:
I've left all the working commits there for now so the steps can be reviewed but I'm happy to squash these too when ready.

philderbeast commented:
Do we use "libonly" to mean tests?

```cabal
-- project-cabal/pkgs/tests.config
packages: Cabal-QuickCheck
packages: Cabal-described
packages: Cabal-tests
packages: Cabal-tree-diff
packages: cabal-testsuite
```

Mikolaj commented:
> Do we use "libonly" to mean tests?

That rings a bell, though you haven't linked to a particular place in code. I think it's a flag that causes tests to test only Cabal-the-library or something like that. CI uses it, IIRC, in one of its phases.

Mikolaj commented:
BTW, I think the separate commits help a lot to understand what is going on, so I'd leave them be.

I hope all the project files are actually used and not dead code themselves...

philderbeast commented:
@Mikolaj I saw "libonly" in the names of two projects:

- `cabal.project.libonly`
- `cabal.project.validate.libonly`

The packages in each look to be for testing:

```cabal
-- project-cabal/pkgs/tests.config
packages:
    Cabal-QuickCheck
  , Cabal-described
  , Cabal-tests
  , Cabal-tree-diff
  , cabal-testsuite
```

Mikolaj commented:
Oh, right, and the project files only contain packages relevant to Cabal-the-library, so it figures.

ulysses4ever commented:
> I hope all the project files are actually used and not dead code themselves...

I'm afraid many of them are dead wood. Most used are cabal.project and cabal.project.validate — locally and in CI respectively. The rest should be grepped and if nothing comes out could be removed imo. One exception is cabal.project.release, which should be checked against the wiki page on making a release (if not mentioned there, could be removed I think).

Kleidukos commented:
I used cabal.project.release to produce FreeBSD bindists for the 3.10.2 series. I hope it was the correct choice. :')

ulysses4ever commented:
Good! Then this is the third one we know is in use. The rest is still under a big question mark.

Mikolaj commented:
Right, some may be used by github CI and others by gitlab CI (the gitlab CI config is in our repo, too).

philderbeast commented:
The only project not referenced is `cabal.project.coverage` and this was the one that wouldn't parse. I'll delete it.

## `cabal.project.buildinfo`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.buildinfo'
./Makefile:63:	cabal build --builddir=dist-newstyle-bi --project-file=cabal.project.buildinfo buildinfo-reference-generator
```

## `cabal.project.coverage`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.coverage'
```

## `cabal.project.doctest`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.doctest'
./Makefile:88:	cabal repl --with-ghc=doctest --build-depends=QuickCheck --build-depends=template-haskell --repl-options="-w" --project-file="cabal.project.doctest" Cabal-syntax
./Makefile:89:	cabal repl --with-ghc=doctest --build-depends=QuickCheck --build-depends=template-haskell --repl-options="-w" --project-file="cabal.project.doctest" Cabal
```

## `cabal.project.libonly`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.libonly'
./Makefile:82:# file from cabal/doctest. This is easy: we just select a project file with no allow-newer (e.g. cabal.project.libonly).
```

## `cabal.project.meta`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.meta'
./Makefile:42:	cabal run --builddir=dist-newstyle-meta --project-file=cabal.project.meta gen-spdx -- templates/SPDX.LicenseId.template.hs $(SPDX_LICENSE_VERSIONS:%=license-list-data/licenses-%.json) $(SPDX_LICENSE_HS)
./Makefile:45:	cabal run --builddir=dist-newstyle-meta --project-file=cabal.project.meta gen-spdx-exc -- templates/SPDX.LicenseExceptionId.template.hs $(SPDX_LICENSE_VERSIONS:%=license-list-data/exceptions-%.json) $(SPDX_EXCEPTION_HS)
./Makefile:55:	cabal run --builddir=dist-newstyle-meta --project-file=cabal.project.meta gen-cabal-macros -- $< $@
./Makefile:58:	cabal run --builddir=dist-newstyle-meta --project-file=cabal.project.meta gen-paths-module -- $< $@
./Makefile:68:	find Cabal-syntax/src Cabal/src cabal-install/src -type f -name '*.hs' | xargs cabal run --builddir=dist-newstyle-meta --project-file=cabal.project.meta analyse-imports --
./Makefile:156:	cabal run --builddir=dist-newstyle-meta --project-file=cabal.project.meta gen-validate-dockerfile -- $* $< $@
```

## `cabal.project.release`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.release'
./README.md:66:    cabal install --project-file=cabal.project.release cabal-install
./Makefile:194:	cabal build --project=cabal.project.release --with-compiler=ghc-$* --dry-run cabal-install:exe:cabal
./.gitlab/ci.sh:57:    echo 'constraints: lukko -ofd-locking' >> cabal.project.release.local
./.gitlab/ci.sh:64:    --project-file=cabal.project.release
./CONTRIBUTING.md:15:cabal build --project-file=cabal.project.release cabal
```

## `cabal.project.validate`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.validate(?!\.libonly)' --perl-regexp
./cabal-testsuite/README.md:40:cabal repl --with-ghc=doctest --build-depends=QuickCheck --build-depends=template-haskell --repl-options="-w" --project-file="cabal.project.validate" Cabal-syntax
./validate.sh:303:    PROJECTFILE=cabal.project.validate
./project-cabal/ghc-latest.config:6:--   - cabal.project.validate (Cabal CI),
./.github/workflows/validate.yml:132:          CABAL_EXEC=$(cabal list-bin --builddir=dist-newstyle-validate-ghc-${{ matrix.ghc }} --project-file=cabal.project.validate cabal-install:exe:cabal)
```

## `cabal.project.validate.libonly`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.validate.libonly'
./validate.sh:301:    PROJECTFILE=cabal.project.validate.libonly
```

## `cabal.project.weeder`
```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.weeder'
./Makefile:182:	cabal build all --project-file=cabal.project.weeder
```

ulysses4ever commented:
Thank you for the analysis. I expected them to show up in the Makefile. The issue is: I don’t think the Makefile itself is widely used. Just look at all the Dockerfile stuff in the Makefile — it’s horrible. It would be nice to clean up the Makefile too.

Let’s look closer at your results:

- I’m pretty sure we don’t use weeded yet (it’d be nice to do it, and of course there is a ticket for it, but we currently don’t), so that one could go.
- cabal.project.libonly looks like a false positive that could go (it’s mentioned in a comment only)
- cabal.project.buildinfo is dubious
- *.meta is mentioned many times in the Makefile but I don’t know that anyone uses it except maybe CI: there is a Meta job in it; someone needs to double check whether CI calls the targets of the Makefile that has to do with this project file.

More radically, I propose to cut all the project files and Makefile targets that
- neither exercised by our CI (GitHub+gitlab)
- nor referenced from the wiki page on Making a release.
Note that to check this condition you have to build a transitive closure: CI can mention a Makefile target that mentions a project file, and so you should preserve the project file…

Putting it positively, we need to make sure that CI uses the Makefile as much as possible, but everything that is not used in CI should be assumed to bit rot and removed. I think there are rare exceptions from this rule that sometimes come from @andreasabel who afair use some non-CI targets in the Makefile. Maybe from @Kleidukos too.

Kleidukos commented:
Yeah I use the makefile for styling.

ulysses4ever commented:
Oh, right, the `style` (and related) target is the recommended way to do it. So, I patch the rule above:

> More radically, I propose to cut all the project files and Makefile targets that
> - neither exercised by our CI (GitHub+gitlab),
> - nor referenced from the wiki page on Making a release,
> - **nor mentioned in CONTRIBUTING.md**

philderbeast commented:
I tried `make weeder` on the master branch at 52d1fb6fc32516e2260f0aecffcb84ded67c4867:

```
$ make weeder
cabal build all --project-file=cabal.project.weeder
Resolving dependencies...
Error: [Cabal-7107]
Could not resolve dependencies:
[__0] trying: cabal-install-3.11.0.0 (user goal)
[__1] next goal: cabal-install-solver (dependency of cabal-install)
[__1] rejecting: cabal-install-solver-3.10.2.1 (conflict: cabal-install => cabal-install-solver^>=3.11)
[__1] skipping: cabal-install-solver-3.10.1.0, cabal-install-solver-3.8.1.0 (has the same characteristics that caused the previous version to fail: excluded by constraint '^>=3.11' from 'cabal-install')
[__1] fail (backjumping, conflict set: cabal-install, cabal-install-solver)
After searching the rest of the dependency tree exhaustively, these were the goals I've had most trouble fulfilling: cabal-install, cabal-install-solver

make: *** [Makefile:182: weeder] Error 1
```

The `cabal-testsuite` `.hie` files will be a [problem for weeder](https://github.com/haskell/cabal/issues/9121#issuecomment-1870634944) and with `weeder >= 2.6` the configuration has changed to TOML. I'll go ahead and remove weeder on this pull request. It can be added back with #9121.

philderbeast commented:
I've updated the Makefile recipe for `doc/buildinfo-fields-reference.rst` instead of using a phony target. Some fixes will be needed to get the users guide to build.

> /home/runner/work/cabal/cabal/doc/buildinfo-fields-reference.rst:516: WARNING: cabal:pkg-field reference target not found: virtual-modules

philderbeast commented:
An interesting effect of this change is finding tests that implicitly pick up the `cabal.project` of cabal itself. That is not what we want is it? Any suggestions on how to isolate tests from our own project?

philderbeast commented:
@Mikolaj I would hope that tests didn't have an inadvertent dependency on something in the root project. Likely they didn't as CI completed green.

ulysses4ever commented:
CI wouldn't help if tests depend on something in the root project file because the file is present in CI as well? I think it's the right thing to do to try to isolate tests from the root project file even if chances it will blow up are small. 

The way to achieve isolation may be debatable. I was thinking our test framework could be smart about inserting --ignore-project when there's no test-local project file. But I don't like smart solutions. The one presented in the respective PR looks simpler, so I say let's KISS…

philderbeast commented:
Is `cabal.project.release` used in cutting a release? I can't see mention of it in the [release wiki article](https://github.com/haskell/cabal/wiki/Making-a-release).

ulysses4ever commented:
I think it does: bindists are build in GitLab CI, and here's the relevant line:

https://github.com/haskell/cabal/blob/49e3d183c0a9ac16da8d4afacc6b8a274ee67608/.gitlab/ci.sh#L64

Generally, try to grep through the repo: sometimes you'll find curious things... :-)

philderbeast commented:
Do we make releases using a project that fails to check that test suites build?

https://github.com/haskell/cabal/blob/49e3d183c0a9ac16da8d4afacc6b8a274ee67608/cabal.project.release#L1-L8

On the master branch at 49e3d183c0a9ac16da8d4afacc6b8a274ee67608:

```
$ cabal build all \
  --enable-tests --enable-benchmarks --dry-run \
  --project-file=cabal.project.release \
  --index-state="hackage.haskell.org HEAD"
Resolving dependencies...
Error: [Cabal-7107]
Could not resolve dependencies:
[__0] trying: cabal-install-3.11.0.0 (user goal)
[__1] trying: cabal-install:*test
[__2] unknown package: Cabal-described (dependency of cabal-install *test)
[__2] fail (backjumping, conflict set: Cabal-described, cabal-install, cabal-install:test)
After searching the rest of the dependency tree exhaustively,
these were the goals I've had most trouble fulfilling: cabal-install, cabal-install:test, Cabal-described
```

philderbeast commented:
What versions of `cabal-install` do we need to support for building cabal? I know that the way imports work changed between `cabal-3.8` and `cabal-3.10`. I had to work around this in Updo, for instance:

https://github.com/search?q=repo%3Acabalism%2Fupdo%20CABAL_RELATIVITY&type=code

alt-romes commented on cabal.project.validate.libonly:
```diff
@@ -1,28 +1,11 @@
-
--- https://github.com/haskell-hvr/cryptohash-sha256/issues/12
-allow-newer: cryptohash-sha256:base
+-- REVIEW: Is the exception for cryptohash-sha256 still needed?
+-- SEE: https://github.com/haskell-hvr/cryptohash-sha256/issues/12
+allow-newer: cryptohash-sha256:base
```
Probably not. Would it be simple to check?

philderbeast replied:
Only needed for `ghc-9.0.*` and `base-4.15.*`, from [ghc-9.0 wiki](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.0#withforeignptr).

philderbeast replied:
But any other library could be likewise affected. Is `ghc-9.0.*` as a whole deprecated?

alt-romes reviewed (approved):
Very good! This is much better than now.

Note: you've included some commits from #9600 so be sure to rebase on that after it lands.

There are also some changes that appear to be fixing ghc 9.8 warnings. I think it is fine for them to be merged in this PR as separate commits, but it might also have been an oversight on your part that included commits from a separate PR.

Thanks!

Mikolaj commented:
> What versions of cabal-install do we need to support for building cabal? 

I'm apparently missing the context. In general, any recent (a couple of major versions back) cabal-install should be buildable with any other, but we won't resurrect old release branches just to fix it, if that's not the case.

fgaz reviewed (commented):
Thanks for the cleanup! The granular commits made it easy to review. Looks good.

A few minor comments:

Mikolaj commented:
CI is broken due to the yesterday's github outage. Let me restart it.

Mikolaj commented:
Let me take the liberty and rebase on master that just got a fix to CI.

Mikolaj commented:
@mergify rebase

ulysses4ever commented:
The above message from the bot is a cryptic way of saying that there are rebase conflicts that it can't solve automatically; hence, a human-driven rebase is needed...

philderbeast commented:
The rebase went through cleanly by itself.

```
$ git rebase master
warning: skipped previously applied commit acb8ffebc
warning: skipped previously applied commit 0f98cbb36
hint: use --reapply-cherry-picks to include skipped commits
hint: Disable this message with "git config advice.skippedCherryPicks false"
dropping f8057fb515e3c78c85880ba31806e91239b460eb Satisfy -Wunused-imports in GenSPDX -- patch contents already upstream
Successfully rebased and updated refs/heads/import/pkg-groups.
```

ulysses4ever commented:
Bots, you can't trust them…

philderbeast commented:
Could someone please have a look at this failure, from [Validate old ghcs 7.10.3](https://github.com/haskell/cabal/actions/runs/7547242267/job/20552913273?pr=9565)?

```
Run sh validate.sh -j 2 -v --lib-only -s lib-suite-extras --extra-hc /opt/ghc/7.10.3/bin/ghc-7.10.3
  
Error: cabal: Internal error in target matching: could not make an unambiguous
fully qualified target selector for 'cabal-testsuite:exe:cabal-tests'.
We made the target 'cabal-testsuite:exe:cabal-tests' (unknown-component) that
was expected to be unambiguous but matches the following targets:
'cabal-testsuite:exe:cabal-tests', matching:
- cabal-testsuite:exe:cabal-tests (unknown-component)
- cabal-testsuite:exe:cabal-tests (unknown-file)
':pkg:cabal-testsuite:exe:cabal-tests', matching:
- cabal-testsuite:exe:cabal-tests (unknown-component)
Note: Cabal expects to be able to make a single fully qualified name for a
target or provide a more specific error. Our failure to do so is a bug in
cabal. Tracking issue: https://github.com/haskell/cabal/issues/8684
Hint: this may be caused by trying to build a package that exists in the
project directory but is missing from the 'packages' stanza in your cabal
project file.
```

There were other earlier GHC versions that were failing but I thought I might be able to test this locally with `ghc-7.10.3` as ghcup lists that as earliest available GHC version. Unfortunately it is failing to install on this ubuntu 23.10 box. Here's the end of `~/.ghcup/logs/ghc-make.log`;

```
Installing library in
~/.ghcup/.../ghc/7.10.3/lib/ghc-7.10.3/ghcpr_8TmvWUcS1U1IKHT0levwg3
gmake[1]: *** [ghc.mk:924: install_packages] Segmentation fault (core dumped)
gmake: *** [Makefile:24: install] Error 2
```

andreabedini commented:
> We have a lot of projects and they each have packages. 

This risks to be misleading: we have many cabal.project files, because we want to compile/run/test our packages in different ways and/or under different conditions.

I really appreciate the analysis of Makefile targets and cabal.project but I am not sure about the reorganisation of the cabal.project files.

I do understand the desire to not repeat the same information everywhere but some changes seems to be unnecessary and unconventional. E.g. the grouping of packages into `project-cabal/pkgs/cabal.config` and `project-cabal/pkgs/install.config seems to be arbitrary. Also why introducing a new cabal.project file for a single `package:` line?

Most likely you are following a pattern here, but I don't know what the pattern is :-)

philderbeast commented:
Did you see the `./project-cabal/README.md` @andreabedini?

philderbeast commented:
> Also why introducing a new cabal.project file for a single package: line?

@andreabedini which project is that? Are you referring to the two singleton package groups?

- `project-cabal/pkgs/buildinfo.config`
- `project-cabal/pkgs/integration-tests.config`

Would you like me to change `project-cabal/pkgs.config` to include those packages directly and drop their package groups?

```diff
$ git diff
diff --git a/project-cabal/pkgs.config b/project-cabal/pkgs.config
index 04b64017f..fc2fdda83 100644
--- a/project-cabal/pkgs.config
+++ b/project-cabal/pkgs.config
@@ -1,6 +1,8 @@
 import: pkgs/cabal.config
 import: pkgs/install.config
-import: pkgs/buildinfo.config
 import: pkgs/tests.config
-import: pkgs/integration-tests.config
-import: pkgs/benchmarks.config
\ No newline at end of file
+import: pkgs/benchmarks.config
+
+packages:
+    buildinfo-reference-generator
+  , cabal-testsuite
\ No newline at end of file
diff --git a/project-cabal/pkgs/buildinfo.config b/project-cabal/pkgs/buildinfo.config
deleted file mode 100644
index 75a899d1c..000000000
--- a/project-cabal/pkgs/buildinfo.config
+++ /dev/null
@@ -1 +0,0 @@
-packages: buildinfo-reference-generator
\ No newline at end of file
diff --git a/project-cabal/pkgs/integration-tests.config b/project-cabal/pkgs/integration-tests.config
deleted file mode 100644
index f047e323f..000000000
--- a/project-cabal/pkgs/integration-tests.config
+++ /dev/null
@@ -1 +0,0 @@
-packages: cabal-testsuite
\ No newline at end of file
```

I've changed and deleted projects but not added any in the root.

![image](https://github.com/haskell/cabal/assets/633283/cc85277c-9c7a-4357-b436-a5ea3f9201ba)

I added one project at [cabal-testsuite/cabal.project](https://github.com/haskell/cabal/pull/9565/files#diff-1f3cdf13b6838853cdbcad9f6560a03d540e6a45373f75b998bc27488ef59c2c) with a comment explaining that it is there to intercept project probing from the tests.

philderbeast commented on project-cabal/pkgs/benchmarks.config:
```diff
@@ -1,2 +1 @@
-packages: cabal-benchmarks
-packages: solver-benchmarks
+packages: cabal-benchmarks, solver-benchmarks
```
Thanks for explaining why the verbose repetition was there in the first place. We don't add or remove packages very often do we?

philderbeast commented on cabal.project.validate:
```diff
@@ -6,17 +9,13 @@ import: project-cabal/pkgs/benchmarks.config
   ghc-options: -Werror
 package cabal-install
-  ghc-options: -Werror
+  ghc-options: -Werror
```
Yikes, sorry. I've added back these newlines to the projects.

andreabedini commented on solver-benchmarks/HackageBenchmark.hs:
```diff
@@ -187,7 +185,7 @@ hackageBenchmarkMain = do
             putStrLn $ "Obtaining the package list (using " ++ argCabal1 ++ ") ..."
             list <- readProcess argCabal1 ["list", "--simple-output"] ""
-            return $ nub [mkPackageName $ head (words line) | line <- lines list]
+            return . nub $ mkPackageName <$> catMaybes [listToMaybe (words line) | line <- lines list]
```
This change is unrelated and I happen to disagree with it. I understand the previous version as "a package name from the first word of each line". While I find the use of listToMaybe/catMaybes in the new version less clear.

Is the goal to avoid `head` because it is a partial function? Would something like 
`[ mkPackageName n | n:_ <- map words (lines line) ]` work?

philderbeast replied:
The change is related. If I revert it, here's the problem it fixes;

```
$ cabal build solver-benchmarks --enable-tests --enable-benchmarks
...
Resolving dependencies...
Build profile: -w ghc-9.8.1 -O1
In order, the following will be built (use -v for more details):
 - Cabal-syntax-3.11.0.0 (lib) (configuration changed)
 - solver-benchmarks-3 (lib) (configuration changed)
 - solver-benchmarks-3 (test:unit-tests) (configuration changed)
 - solver-benchmarks-3 (exe:hackage-benchmark) (configuration changed)
Configuring library for Cabal-syntax-3.11.0.0...
Preprocessing library for Cabal-syntax-3.11.0.0...
Building library for Cabal-syntax-3.11.0.0...
Configuring library for solver-benchmarks-3...
Preprocessing library for solver-benchmarks-3...
Building library for solver-benchmarks-3...
[1 of 1] Compiling HackageBenchmark

HackageBenchmark.hs:188:43: error: [GHC-63394] [-Wx-partial, -Werror=x-partial]
    In the use of ‘head’
    (imported from Prelude, but defined in GHC.List):
    "This is a partial function, it throws an error on empty lists.
    Use pattern matching or Data.List.uncons instead.
    Consider refactoring to use Data.List.NonEmpty."
    |
188 |             return $ nub [mkPackageName $ head (words line) | line <- lines list]
    |                                           ^^^^

HackageBenchmark.hs:340:44: error: [GHC-63394] [-Wx-partial, -Werror=x-partial]
    In the use of ‘head’
    (imported from Prelude, but defined in GHC.List):
    "This is a partial function, it throws an error on empty lists.
    Use pattern matching or Data.List.uncons instead.
    Consider refactoring to use Data.List.NonEmpty."
    |
340 |   | allEqual rs                          = head rs
    |                                            ^^^^
Error: [Cabal-7125]
Failed to build solver-benchmarks-3
(which is required by test:unit-tests from solver-benchmarks-3 and exe:hackage-benchmark
from solver-benchmarks-3).
```

philderbeast replied:
Thanks for your suggestion that I like a lot.

I'm a fan of list comprehensions but haven't been using partial matches with them. Wow, have I been missing out!

philderbeast replied:
I've gone ahead an used a partial pattern for filtering.

andreabedini reviewed (dismissed):
@philderbeast

> Did you see the ./project-cabal/README.md @andreabedini? There's a table there showing which projects import which package groups.

I have seen it and I read it but what are "package groups"? You are referring to conventions that are not widespread.

I am very sorry but I don't understand what we want to achieve in this PR. The description basically only says "use package groups" but it does not say "why".

Do you see a problem in the presence of many cabal.project files? That's fair. I see that you have already put work into understanding where each one is used. Maybe we can remove the need for some of them. Project configuration options can also passed on the command line so

```
-- cabal.project
package: a b c

-- cabal.project.variant
import cabal.project
variant: a
```
```
cabal build
cabal build --project-file cabal.project.variant
```

could simplified to

```
-- cabal.project
package: a b c
```
```
cabal build
cabal build --variant=a
```

What you propose is an interesting machinery but I don't see why we should adopt it.

philderbeast commented:
> I have seen it and I read it but what are "package groups"? You are referring to conventions that are not widespread.

@andreabedini by "package group" I mean a `.config` file that only has a `packages` stanza (imported into a project).

philderbeast commented:
> I am very sorry but I don't understand what we want to achieve in this PR.

This is a project maintenance change. I'll raise a motivating issue with more details on what I've found.

philderbeast commented:
## Project comparisons

### Before this Change

```diff
$ diff cabal.project cabal.project.validate
2a3
> packages: Cabal-syntax/
5d5
< packages: Cabal-syntax/
7d6
< packages: cabal-install-solver/
10,11c9
< tests: True
< 
---
> packages: cabal-install-solver/
13c11
< packages: Cabal-tree-diff/
---
> packages: Cabal-tree-diff
15,16c13,14
< packages: Cabal-tests/
< packages: cabal-benchmarks/
---
> packages: Cabal-tests
> packages: cabal-benchmarks
18c16
< optional-packages: ./vendored/*/*.cabal
---
> tests: True
23a22,23
> write-ghc-environment-files: never
> 
25a26,34
> 
> package Cabal-syntax
>   ghc-options: -Werror
> package Cabal
>   ghc-options: -Werror
> package cabal-testsuite
>   ghc-options: -Werror
> package cabal-install
>   ghc-options: -Werror

$ diff cabal.project cabal.project.libonly 
1,10c1
< import: cabal.project.latest-ghc
< 
< packages: Cabal/
< packages: cabal-testsuite/
< packages: Cabal-syntax/
< packages: cabal-install/
< packages: cabal-install-solver/
< packages: solver-benchmarks/
< 
< tests: True
---
> packages: Cabal-syntax/ Cabal/ cabal-testsuite/
13c4
< packages: Cabal-tree-diff/
---
> packages: Cabal-tree-diff
15,16c6
< packages: Cabal-tests/
< packages: cabal-benchmarks/
---
> packages: Cabal-tests
18c8
< optional-packages: ./vendored/*/*.cabal
---
> tests: True
20,22c10,11
< -- avoiding extra dependencies
< constraints: rere -rere-cfg
< constraints: these -assoc
---
> -- Uncomment to allow picking up extra local unpacked deps:
> --optional-packages: */

$ diff cabal.project.libonly cabal.project.validate.libonly 
1,2c1,3
< packages: Cabal-syntax/ Cabal/ cabal-testsuite/
< 
---
> packages: Cabal-syntax/
> packages: Cabal/
> packages: cabal-testsuite/
10,11c11,15
< -- Uncomment to allow picking up extra local unpacked deps:
< --optional-packages: */
---
> write-ghc-environment-files: never
> 
> -- avoiding extra dependencies
> constraints: rere -rere-cfg
> constraints: these -assoc
14a19,28
> 
> package Cabal-syntax
>   ghc-options: -Werror
> package Cabal
>   ghc-options: -Werror
> package cabal-testsuite
>   ghc-options: -Werror
> 
> -- https://github.com/haskell-hvr/cryptohash-sha256/issues/12
> allow-newer: cryptohash-sha256:base
```

### After this Change

```diff
$ diff cabal.project cabal.project.validate
6a7
> write-ghc-environment-files: never

$ diff cabal.project cabal.project.libonly
2,4c2,3
< import: project-cabal/ghc-latest.config
< import: project-cabal/pkgs.config
< import: project-cabal/constraints.config
---
> import: project-cabal/pkgs/cabal.config
> import: project-cabal/pkgs/tests.config

$ diff cabal.project.libonly cabal.project.validate.libonly
3a4
> import: project-cabal/constraints.config
5a7,11
> write-ghc-environment-files: never
> 
> -- REVIEW: Is the exception for cryptohash-sha256 still needed?
> -- SEE: https://github.com/haskell-hvr/cryptohash-sha256/issues/12
> allow-newer: cryptohash-sha256:base
```

philderbeast commented:
Rerunning the grep, @ulysses4ever it looks like I could get rid of `cabal.project.libonly` too;

```
$ grep -rwn --exclude-dir=".git" . -e 'cabal.project.libonly'
./Makefile:88:# file from cabal/doctest. This is easy: we just select a project file with no allow-newer (e.g. cabal.project.libonly).
./project-cabal/README.md:9:├── cabal.project.libonly
```

https://github.com/haskell/cabal/blob/c32dcfaf182953d302cae28c1d6d5abdaed5bea3/Makefile#L84-L88

We might enforce `--allow-newer=False` with this change;

```diff
$ git diff
diff --git a/Makefile b/Makefile
index 685d88a34..061a4efe9 100644
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,7 @@
 
 CABALBUILD := cabal build
 CABALRUN   := cabal run
-DOCTEST    := cabal repl --with-ghc=doctest --repl-options="-w" --ghc-options="-Wwarn"
+DOCTEST    := cabal repl --with-ghc=doctest --repl-options="-w" --ghc-options="-Wwarn" --allow-newer=False
 
 # default rules
```

That would catch a failing CI build for doctests early.

ulysses4ever commented:
Yes, thank you, removing .libonly seems no-brainer if it's not used in the CI. I thought it was used because we have "lib-only" steps in CI. But I trust you and grep.

While I'm here, on the general note of restructuring the project files...

- The change is big, and will be disruptive in several ways, some of which are not yet known, but first of all mentally: Cabal devs are used to the old structure, and the new one is pretty far apart, at least superficially.

- The disruptiveness will be compounded by the fact that the new structure doesn't follow widely-known patterns of structuring Haskell projects. This alone could kill it for many more conservative project maintainers (in any project).

- As a result, I believe such a change requires solid motivation and documentation.

- More than that. I always lean on the "let's try new stuff" side, so I'm open for most non-breaking changes, but you'll have to convince a non-negligible number of interested parties. As such, I don't think any 2 approvals should be enough: there should be no unresolved points in this thread. But this is my personal feeling, not a rule.

philderbeast commented:
@ulysses4ever do you think adding `--allow-newer=False` is even needed now that #8732 has merged.

ulysses4ever commented:
Until we release that feature, I guess so?..

Kleidukos reviewed (approved):
I agree with this change, thank you very much @philderbeast

philderbeast commented:
> Project configuration options can also passed on the command line so

Thanks @andreabedini. This works because command line options have priority, don't they? The docs don't mention that:

https://github.com/haskell/cabal/blob/1a8b93c1f1717dd454ff50546f890da734bbf45d/doc/cabal-project-description-file.rst#L23-L33

andreabedini commented:
> Thanks @andreabedini. This works because command line options have priority, don't they? The docs don't mention that (with projects - maybe it does elsewhere?):

Yes, that's right. I described how it works [in this comment](https://github.com/haskell/cabal/blob/ae3c40a20bf98870488e3b40fc4495009b026e33/cabal-install/src/Distribution/Client/ProjectPlanning.hs#L298) in `ProjectPlanning`.

I guess the relevent paragraph in the docs is [5.2.7.1. cabal build](https://cabal.readthedocs.io/en/stable/cabal-commands.html#cabal-build) but I agree that it could be more explicit.

> Beyond a list of targets, cabal build accepts all the flags that cabal configure takes. Most of these flags are only taken into consideration when building local packages; however, some flags may cause extra store packages to be built (for example, --enable-profiling will automatically make sure profiling libraries for all transitive dependencies are built and installed.)

philderbeast commented on cabal.project:
```diff
@@ -1,6 +1,4 @@
-program-options
-  ghc-options: -fno-ignore-asserts
-
+import: project-cabal/ghc-options.config
```
@fgaz what's your opinion on the `cabal.project` name? That's for dev too isn't it?

philderbeast commented:
> Do you see a problem in the presence of many cabal.project files? That's fair. I see that you have already put work into understanding where each one is used. Maybe we can remove the need for some of them. Project configuration options can also passed on the command line

@andreabedini I dare say we could get by with a single `cabal.project` and use command line flags. I'm leaning that way (and didn't see this option immediately). However, I think this is too radical to transition to in a single step so went with a gentler tidying up.

andreabedini commented on cabal.project.validate.libonly:
```diff
@@ -1,28 +1,12 @@
-  ghc-options: -Werror
-
--- https://github.com/haskell-hvr/cryptohash-sha256/issues/12
+-- REVIEW: Is the exception for cryptohash-sha256 still needed?
+-- SEE: https://github.com/haskell-hvr/cryptohash-sha256/issues/12
```
Since the ticket is closed, it looks like it isn't. Just remove it and see what the CI says.

philderbeast replied:
Trying now. Fingers crossed.

philderbeast replied:
That worked, CI is all green.

Mikolaj commented:
> @andreabedini I dare say we could get by with a single `cabal.project` and use command line flags. I'm leaning that way (and didn't see this option immediately). However, I think this is too radical to transition to in a single step so went with a gentler tidying up.

We did not discuss the design enough in the issue ticket for this PR, so this is more of an exploratory PR but, since it turns out to be better than status quo, I'm for merging it as is. However, afterwards, I feel a design discussion or a new design demonstration in the form of a new PR would be proper at any time, even without a user/dev experience gathering period, if anybody feels so inclined. The lack of ceremony cuts both ways. Thank you, everybody.

philderbeast commented:
> BTW, I think the separate commits help a lot to understand what is going on, so I'd leave them be.

@Mikolaj it is OK to squash down to one or a few commits now that the reviews are in, isn't it?

Mikolaj commented:
> @Mikolaj it is OK to squash down to one or a few commits now that the reviews are in, isn't it?

It's your call and, as recommendations go, squash is the default.

philderbeast commented:
@Mikolaj, I got the 40 commits down to 6;

- 1 for importing packages
- 1 for deleting projects
- 1 for building all local packages with `-Werror`
- 1 for the cabal-testsuite/cabal.project intercepting project (so that cabal's own project is not picked up by tests)
- a couple more for minor isolated things

philderbeast commented:
Label merge+no rebase is necessary when the pull request is from an organisation.

philderbeast commented:
@Mikolaj I don't know what is going on with this pull request. It has no conflicts with the base branch, is approved with a `merge+no rebase` label applied 5 days ago. I thought the wait was 2 days and that it would have merged by now.

ulysses4ever commented:
@Mergify refresh

ulysses4ever commented:
@mergify refresh

ulysses4ever commented:
Looking at the Checks tab, it looks like the bot doesn't see enough approvals. I have no idea why. Maybe merging manually is the only way forward...

geekosaur commented:
Does mergify care about "changes requested"? It looks to me like there are two reviews in that state.

philderbeast commented:
> Does mergify care about "changes requested"? It looks to me like there are two reviews in that state.

Easy to miss when they're hidden behind a pleat;

![image](https://github.com/haskell/cabal/assets/633283/b1633d01-9728-4e4f-be57-e4f774ce98b7)

philderbeast commented:
@fgaz could you have a look at your review notes again please?

Mikolaj commented:
I don't see red review marks ("changes requested"), so I don't think the review are blocking it. Let me try to insist.

Mikolaj commented:
@mergify dequeue

Mikolaj commented:
@mergify queue

Mikolaj commented:
I think mergify lost track of the `approved-reviews-by` number.

Mikolaj reviewed (changes requested):
Testing.

Mikolaj reviewed (approved):
Testing 2.

Mikolaj commented:
@philderbeast: could you kindly rebase manually and then manually merge with no-fast-forward (e.g., via the "merge pull request" button)? Thank you!

philderbeast commented:
@Mikolaj thanks very much for debugging mergify. I've made the merge.

Mikolaj commented:
Well done! :)

philderbeast commented:
I validated master locally to be sure after the merge and it is all good (after the 15 min wait).

```
$ ./validate.sh
...
!!! Validation took 908 seconds.
```

ulysses4ever commented:
@mergify backport 3.12
